File: Problems/438_Find_All_Anagrams_in_a_String.cpp
#include<string>
#include<vector>
using namespace std;
class Solution {
public:
    vector<int> findAnagrams(string s, string p) {
        vector<int> res;
        int lenS = s.length();
        int lenP = p.length();
        if(lenS<lenP) return res;

        vector<int> alphaS(26, 0);
        for(int i=0; i<lenP; i++) alphaS[s[i]-'a']++;

        vector<int>alphaP(26, 0);
        for(auto c: p) alphaP[c-'a']++;

        if(alphaS==alphaP) res.push_back(0);
        for(int i=1; i<lenS-lenP+1; i++){
            alphaS[s[i-1]-'a'] --;
            alphaS[s[i-1+lenP]-'a'] ++;
            if(alphaS==alphaP) res.push_back(i);
        }
        return res;
    }
};

File: Problems/438_Find_All_Anagrams_in_a_String.py
from typing import List

class Solution:
    def findAnagrams(self, s: str, p: str) -> List[int]:
        res = []
        lengthS = len(s)
        lengthP = len(p)
        if lengthS<lengthP:
            return res

        alphaP = [0]*26
        for c in p:
            alphaP[ord(c)-ord('a')] += 1
            
        alphaS = [0]*26
        for i in range(lengthP):
            alphaS[ord(s[i])-ord('a')] += 1
        
        if alphaS==alphaP:
            res.append(0)
        for i in range(1, lengthS-lengthP+1):
            alphaS[ord(s[i-1])-ord('a')]-=1
            alphaS[ord(s[i+lengthP-1])-ord('a')] += 1
            if alphaS==alphaP:
                res.append(i)
        return res

if __name__ == '__main__':
    sol = Solution()
    s = "abab"
    p = "ab"
    print(sol.findAnagrams(s, p))